src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -255,21 +309,105 @@ cdef class AioServer:
         return self._server.add_http2_port(address,
                                           server_credentials._credentials)
 
+    async def _server_main_loop(self,
+                                object server_started):
+        self._server.start(backup_queue=False)
+        cdef RPCState rpc_state
+        server_started.set_result(True)
+
+        while True:
+            # When shutdown begins, no more new connections.
+            if self._status != AIO_SERVER_STATUS_RUNNING:
+                break
+
+            rpc_state = await _server_call_request_call(
+                self._server,
+                self._cq,
+                self._loop)
+
+            self._loop.create_task(_handle_rpc(
+                self._generic_handlers,
+                rpc_state,
+                self._loop))
+
     async def start(self):
         if self._status == AIO_SERVER_STATUS_RUNNING:
             return
         elif self._status != AIO_SERVER_STATUS_READY:
             raise RuntimeError('Server not in ready state')
 
         self._status = AIO_SERVER_STATUS_RUNNING
-        loop = asyncio.get_event_loop()
-        loop.create_task(_server_start(
-            self._server,
-            self._cq,
-            self._generic_handlers,
-        ))
+        cdef object server_started = self._loop.create_future()
+        self._serving_task = self._loop.create_task(self._server_main_loop(server_started))
+        # Needs to explicitly wait for the server to start up.
+        # Otherwise, the actual start time of the server is un-controllable.
+        await server_started
+
+    async def _start_shutting_down(self):
+        """Prepares the server to shutting down (NOT coroutine-safe)."""
+        # Starts the shutdown process.
+        # The shutdown callback won't be called until there is no live RPC.
+        grpc_server_shutdown_and_notify(
+            self._server.c_server,
+            self._cq._cq,
+            self._shutdown_callback_wrapper.c_functor())
 
-    # TODO(https://github.com/grpc/grpc/issues/20668)
-    # Implement Destruction Methods for AsyncIO Server
-    def stop(self, unused_grace):
-        pass
+        # Ensures the serving task (coroutine) exits.
+        try:
+            await self._serving_task
+        except _RequestCallError:
+            pass
+
+    async def shutdown(self, grace):
+        """Gracefully shutdown the C-Core server.
+
+        Application should only call shutdown once.
+
+        Args:
+          grace: An optional float indicating the length of grace period in
+            seconds.
+        """
+        if self._status == AIO_SERVER_STATUS_READY or self._status == AIO_SERVER_STATUS_STOPPED:
+            return
+
+        async with self._shutdown_lock:
+            if self._status == AIO_SERVER_STATUS_RUNNING:
```

Comment:
Correct me if my reasoning is wrong here. The only place where execution is yielded back to the event loop is before and after the call to `self._start_shutting_down()`. So the point of the lock is to ensure that the rest of the program observes the call to `start_shutting_down` and the updates to these values as a single atomic unit. Couldn't this just be achieved by making `_start_shutting_down()` synchronous?